Hermite finite elements - initial pull request

ivweber opened the pull request:
I have created code files to implement Hermite elements capable of enforcing continuity in higher order derivatives across element boundaries on a Cartesian grid, and I now believe that these files are ready to be merged into the deal.II library.

The files polynomials_hermite.h and polynomials_hermite.cc implement a Polynomial class containing the Hermite polynomials.

The files fe_hermite.h, fe_hermite.cc and fe_hermite.inst.in implement a FE_Poly object storing the Hermite polynomials, as well as certain member functions that are needed elsewhere.

The files vector_tools_hermite.h, vector_tools_hermite.templates.h, vector_tools_hermite.cc and vector_tools_hermite.inst.in implement projection functions for use with Hermite elements. These additional functions were found to be necessary due to Hermite elements requiring different treatment of degrees of freedom on the domain boundary.

Some existing files have had minor edits to allow compatibility with Hermite elements, mainly involving the removal of some assert statements that were triggered by the properties of Hermite elements.

Additionally 7 test files have been added to test the implemented functionality for Hermite elements.

- Ivy Weber, 14th January 2022

peterrum commented:
@DanielPaukner Could you give a try to this code.

kronbichler commented:
> @DanielPaukner Could you give a try to this code.

I am not sure what kind of application you are considering, but I anticipate that typical beams will need to provide a suitable mapping that is not yet there in this case (the current version works with `MappingCartesian`). Anyway, I believe or at least hope this PR should lay some foundations that can make the additional work limited.

drwells commented:
This is fantastic! Hermite elements are a commonly requested feature. Even getting this working with Cartesian grids is a nice step forward.

DanielPaukner commented:
Great work!:) And sure, I will give it a try! @kronbichler so far I have used the `MappingFEField` class.

ivweber commented:
I've modified the test files so that log output from the CG solver isn't included in the tested output, it looked as if this might be dependent on the system the tests were running on since errors were coming from a different number of iterations. I've also removed a couple of test cases from the tests where the matrix system was slower to converge or did not completely converge.

peterrum commented on tests/fe/uneven_1d_grid_hermite.cc:
```diff
@@ -0,0 +1,192 @@
+
+  err_sq = std::sqrt(err_sq);
+
+  char fname[50];
+  sprintf(fname, "Cell-1d-Hermite-%d", regularity);
```
`std::string fname = "Cell-1d-Hermite-" + std::to_string(regularity)`

peterrum commented on tests/fe/uneven_1d_grid_hermite.cc:
```diff
@@ -0,0 +1,192 @@
+//
+// ---------------------------------------------------------------------
+
+#include <deal.II/base/function.h>
+#include <deal.II/base/quadrature_lib.h>
+#include <deal.II/base/utilities.h>
+
+#include <deal.II/dofs/dof_accessor.h>
+#include <deal.II/dofs/dof_handler.h>
+#include <deal.II/dofs/dof_tools.h>
+
+#include <deal.II/fe/fe_hermite.h>
+#include <deal.II/fe/fe_interface_values.h>
+#include <deal.II/fe/fe_values.h>
+#include <deal.II/fe/mapping_cartesian.h>
+
+#include <deal.II/grid/grid_generator.h>
+#include <deal.II/grid/tria.h>
+#include <deal.II/grid/tria_accessor.h>
+#include <deal.II/grid/tria_iterator.h>
+
+#include <deal.II/lac/affine_constraints.h>
+#include <deal.II/lac/dynamic_sparsity_pattern.h>
+#include <deal.II/lac/full_matrix.h>
+#include <deal.II/lac/precondition.h>
+#include <deal.II/lac/solver_cg.h>
+#include <deal.II/lac/sparse_matrix.h>
+#include <deal.II/lac/vector.h>
+
+#include <deal.II/numerics/data_out.h>
+#include <deal.II/numerics/matrix_tools.h>
+#include <deal.II/numerics/vector_tools.h>
+
+#include <cmath>
+#include <cstdlib>
+#include <fstream>
+#include <iterator>
+#include <sstream>
+#include <vector>
```
Do yo need so much includes?

peterrum commented on tests/fe/uneven_1d_grid_hermite.cc:
```diff
@@ -0,0 +1,192 @@
+using namespace dealii;
+
+// Define a function to project onto the domain
+class test_poly : public Function<1>
```
```suggestion
class TestPoly : public Function<1>
```

peterrum commented on tests/fe/uneven_1d_grid_hermite.cc:
```diff
@@ -0,0 +1,192 @@
+
+  double err_sq = 0;
+
+  for (auto &cell : dof.active_cell_iterators())
+    {
+      fe_herm.reinit(cell);
+      cell->get_dof_indices(local_to_global);
+      for (const unsigned int q : fe_herm.quadrature_point_indices())
+        {
+          double sol_at_point = 0;
+          for (const unsigned int i : fe_herm.dof_indices())
+            sol_at_point +=
+              fe_herm.shape_value(i, q) * solution(local_to_global[i]);
+
+          sol_at_point -= rhs_func.value(fe_herm.quadrature_point(q));
+          err_sq += sol_at_point * sol_at_point * fe_herm.JxW(q);
+        }
+    }
```
Do yo need to compute this by hand? Cannot you use functions from the `VectorTools` namespace.

peterrum commented on tests/fe/uneven_1d_grid_hermite_laplace.cc:
```diff
@@ -0,0 +1,250 @@
+using namespace dealii;
+
+// Define a function to project onto the domain [-1,1]
+class solution : public Function<1>
```
```suggestion
class Solution : public Function<1>
```
same below

peterrum commented on tests/numerics/wave_solver_hermite.cc:
```diff
@@ -0,0 +1,425 @@
+  MatrixTools::apply_boundary_values(
+    boundary_values, mass_solve, sol_next, temp, true);
+
+  SparseDirectUMFPACK mass_inv;
```
I think you can compile deal.II without UMFPACK support. Either, we exclude the test if UMFPACK is missing or we use a simpler iterative deal.II iterative solver.

peterrum commented on tests/numerics/wave_solver_hermite.cc:
```diff
@@ -0,0 +1,425 @@
+  deallog << "Final time: " << final_time << std::endl;
+  deallog << std::endl;
+
+  deallog << "Error at final time: " << *(--l2_errors.cend()) << std::endl;
```
```suggestion
  deallog << "Error at final time: " << l2_errors.back() << std::endl;
```

peterrum commented on tests/numerics/wave_solver_hermite.cc:
```diff
@@ -0,0 +1,425 @@
+  deallog << std::setprecision(8) << std::fixed;
+  deallog.attach(logfile);
+
+  test_wave_solver<1>(-0.3, 1);
+
+  test_wave_solver<1>(0.0, 1);
+  test_wave_solver<1>(0.0, 2);
+  test_wave_solver<1>(0.0, 3);
+
+  test_wave_solver<2>(0.0, 1);
+  test_wave_solver<2>(0.0, 2);
+  test_wave_solver<2>(0.0, 3);
+
+  test_wave_solver<3>(0.0, 1);
```
I am wondering how long this test is running? If it takes too long, we could consider to decrease the number of timesteps.

peterrum commented on tests/fe/2d_grid_projection_hermite.cc:
```diff
@@ -0,0 +1,223 @@
+
+
+/*
+ * Test case for Hermite on an irregular 1D grid.
+ * <code>FE_Hermite<dim>(reg)<\code> should be able to perfectly represent any
+ * polynomial function up to degree $2 \times reg+1$. If all basis functions
+ * are correctly scaled according to element size, then projecting a
```
I think this text is copy-and-paste. What is the difference to other tests? You can also reference other tests.

peterrum commented on include/deal.II/numerics/vector_tools_hermite.h:
```diff
@@ -0,0 +1,133 @@
+   */
+
+  /**
+   * @p Enumeration representing the type of boundary condition to be enforced.
```
```suggestion
   * Enumeration representing the type of boundary condition to be enforced.
```

peterrum commented on include/deal.II/numerics/vector_tools_hermite.h:
```diff
@@ -0,0 +1,133 @@
+   * to impose boundary conditions such as Neumann conditions directly to reduce
+   * the size of the computational system.
+   */
+  enum HermiteBoundaryType
+  {
+    hermite_dirichlet,
+    hermite_neumann,
+    hermite_2nd_derivative,
+  };
```
I am wondering if this general enough. Would someone to fix multiple components? What about `hermite_5th_derivative`? Wouldn't an approach via `ComponentMask` more flexible?

peterrum replied:
?

peterrum commented on include/deal.II/numerics/vector_tools_hermite.h:
```diff
@@ -0,0 +1,133 @@
+
+
+  /**
+   * Enforces boundary conditions by projecting onto the Hermite finite element
```
```suggestion
   * Enforce boundary conditions by projecting onto the Hermite finite element
```

peterrum commented on include/deal.II/numerics/vector_tools_hermite.h:
```diff
@@ -0,0 +1,133 @@
+   * those degrees of freedom from the boundary system.
+   *
+   * Note that currently only Dirichlet, Neumann and second degree Neumann
+   * conditions are currently supported.
```
```suggestion
   * conditions are supported.
```

peterrum replied:
2x `currently` in the sentence.

peterrum commented on include/deal.II/numerics/vector_tools_hermite.h:
```diff
@@ -0,0 +1,133 @@
+  /**
+   * The following does exactly the same as the above, but uses
+   * hermite_dirichlet as the value
+   * for @p projection_mode.
```
```suggestion
   * for projection_mode.
```
(not a parameter in this function)

peterrum commented on include/deal.II/numerics/vector_tools_hermite.h:
```diff
@@ -0,0 +1,133 @@
+
+
+  /**
+   * The following does exactly the same as the above, but uses
```
The following function is a specialization of the above function for the case that Dirichlet BC should be enforced.

peterrum replied:
```suggestion
The following function is a specialization of the above function for the case that Dirichlet BC should be enforced.
```

peterrum commented on include/deal.II/numerics/vector_tools_hermite.h:
```diff
@@ -0,0 +1,133 @@
+
+
+  /**
+   * Projection function for use with Hermite finite elements. Usually the existing @p project
```
```suggestion
   * Projection function for use with Hermite finite elements. Usually the existing VectorTools::project()
```

peterrum commented on include/deal.II/numerics/vector_tools_hermite.h:
```diff
@@ -0,0 +1,133 @@
+
+  /**
+   * Projection function for use with Hermite finite elements. Usually the existing @p project
+   * function will work with Hermite, however setting any of the flags for
```
Question: have you added asserts there (so that one does not use the function if boundary terms are enabled)?

peterrum commented on include/deal.II/numerics/vector_tools_hermite.h:
```diff
@@ -0,0 +1,133 @@
+  template <int dim, typename VectorType, int spacedim>
+  void
+  project_hermite(
+    const Mapping<dim, spacedim> &                             mapping_h,
+    const DoFHandler<dim, spacedim> &                          dof,
```
```suggestion
    const Mapping<dim, spacedim> &                             mapping,
    const DoFHandler<dim, spacedim> &                          dofhandler,
```

peterrum commented on include/deal.II/numerics/vector_tools_hermite.h:
```diff
@@ -0,0 +1,133 @@
+    const AffineConstraints<typename VectorType::value_type> & constraints,
+    const Quadrature<dim> &                                    quadrature,
+    const Function<spacedim, typename VectorType::value_type> &function,
+    VectorType &                                               vec,
```
In principal, this should work in parallel?

peterrum reviewed (commented):
Let me post what I have till now. I still have three (bigger) files to go: `fe_hermite.h`/`fe_hermite.cc`, and `vector_tools_hermite.templates.h`. 

What I have seen till now looks very good :+1:

peterrum commented:
You will need to add a change-log entry here: https://github.com/dealii/dealii/tree/master/doc/news/changes/major

peterrum commented:
One question: I thought you had to adjust the mapping class. Is this part of this PR?

ivweber commented on include/deal.II/base/polynomials_hermite.h:
```diff
@@ -0,0 +1,195 @@
+   * numer of functions representing shape value and derivatives
+   * at $y=0$ and $y=1$ on the reference interval $y \in [0,1]$.
+   *
+   * Indices $0 \leq j \leq q$ refer to polynomials corresponding
+   * to a non-zero derivative (or shape value for $j=0$) of
+   * order $j$ at $x=0$, and indices $q+1 \leq j \leq 2q+1$
```
I will change q to r and state explicitly that this is the value of the regularity.

ivweber commented on include/deal.II/base/polynomials_hermite.h:
```diff
@@ -0,0 +1,195 @@
+
+
+    /**
+     * This function is used internally to generate a vector of Polynomial
```
This is used by fe_hermite rather than for use by an end user, but you're right that this is badly worded. I will change the description to better reflect its purpose in the code.

peterrum commented on source/fe/fe_hermite.cc:
```diff
@@ -0,0 +1,840 @@
+
+
+  template <int xdim, int xspacedim = xdim, typename xNumber = double>
+  class Rescaler
```
I would move this into an anonymous namespace. Also, I would remove the reference to this class in the documentation of `FEHermtie`, since the it is implementation detail and users do not have to care that the work is outsourced to this class. Furthermore, could you add some documentation to this class even though it is internal.

peterrum commented on source/fe/fe_hermite.cc:
```diff
@@ -0,0 +1,840 @@
+      unsigned int n_q_points;
+      double       cell_extent = 1.0;
+
+      // Check mapping_data is associated with a compatible mapping class
+      if (dynamic_cast<const typename MappingHermite<1, spacedim>::InternalData
+                         *>(&mapping_data) != nullptr)
+        {
+          const typename MappingHermite<1, spacedim>::InternalData *data =
+            dynamic_cast<
+              const typename MappingHermite<1, spacedim>::InternalData *>(
+              &mapping_data);
+          n_q_points  = data->quadrature_points.size();
+          cell_extent = data->cell_extents[0];
+        }
+      else if (dynamic_cast<const typename MappingCartesian<1>::InternalData *>(
+                 &mapping_data) != nullptr)
+        {
+          const typename MappingCartesian<1>::InternalData *data =
+            dynamic_cast<const typename MappingCartesian<1>::InternalData *>(
+              &mapping_data);
+          n_q_points  = data->quadrature_points.size();
+          cell_extent = data->cell_extents[0];
+        }
```
I guess the two cases are identical. If yes, I would remove one and add a case if there is indeed a new mapping class.

peterrum commented on source/fe/fe_hermite.cc:
```diff
@@ -0,0 +1,840 @@
+
+      FullMatrix<double> factors(dofs_per_dim, dofs_per_dim);
+
+      for (unsigned int q = 0; q < q_points; ++q)
+        {
+          double factor_2 = 1.0;
+
+          for (unsigned int d3 = 0, d4 = regularity + 1; d4 < dofs_per_dim;
+               ++d3, ++d4)
+            {
+              double factor_1 = factor_2;
+
+              for (unsigned int d1 = 0, d2 = regularity + 1; d2 < dofs_per_dim;
+                   ++d1, ++d2)
+                {
+                  factors(d1, d3) = factor_1;
+                  factors(d2, d3) = factor_1;
+                  factors(d1, d4) = factor_1;
+                  factors(d2, d4) = factor_1;
+
+                  value_list(l2h[d1 + d3 * dofs_per_dim], q) *= factor_1;
+                  value_list(l2h[d2 + d3 * dofs_per_dim], q) *= factor_1;
+                  value_list(l2h[d1 + d4 * dofs_per_dim], q) *= factor_1;
+                  value_list(l2h[d2 + d4 * dofs_per_dim], q) *= factor_1;
+
+                  factor_1 *= cell_extents[0];
+                }
+
+              factor_2 *= cell_extents[1];
+            }
+        }
```
I know this type of code^^ Could you add some comments that you have unrolled the loop over nodes.

peterrum commented on source/fe/fe_hermite.cc:
```diff
@@ -0,0 +1,840 @@
+              for (unsigned int d1 = 0, d2 = regularity + 1; d2 < dofs_per_dim;
+                   ++d1, ++d2)
+                {
+                  factors(d1, d3) = factor_1;
+                  factors(d2, d3) = factor_1;
+                  factors(d1, d4) = factor_1;
+                  factors(d2, d4) = factor_1;
```
This does not seem to be be used.

peterrum commented on source/fe/fe_hermite.cc:
```diff
@@ -0,0 +1,840 @@
+      else
+        Assert(false, ExcInternalError());
+
+      const unsigned int dofs          = fe_herm.n_dofs_per_cell();
```
```suggestion
      const unsigned int n_dofs_per_cell          = fe_herm.n_dofs_per_cell();

```

ivweber commented on include/deal.II/base/polynomials_hermite.h:
```diff
@@ -0,0 +1,195 @@
+     *
+     * @param regularity
+     */
+    static FullMatrix<double>
+    hermite_to_bernstein_matrix(const unsigned int regularity);
+
+
+    /**
+     * This function generates a vector of coefficients that can be
+     * used by the existing Polynomial class constructor to form
+     * the required objects.
+     */
+    static std::vector<double>
+    hermite_poly_coeffs(const unsigned int regularity,
+                        const unsigned int index);
```
I started investigating the possibility of implementing partial grid refinement with Hermite elements, and the function hermite_to_bernstein_matrix will be needed then so I would prefer to leave it as publicly available for now. I have moved hermite_poly_coeffs into an internal namespace since it wasn't needed in the header file when the constructor is defined in the source code file.

peterrum commented on source/fe/fe_hermite.cc:
```diff
@@ -0,0 +1,840 @@
+
+      const unsigned int dofs          = fe_herm.n_dofs_per_cell();
+      const unsigned int regularity    = fe_herm.get_regularity();
+      const unsigned int dofs_per_dim  = 2 * regularity + 2;
```
```suggestion
      const unsigned int n_dofs_per_dim  = 2 * regularity + 2;

```

peterrum commented on source/fe/fe_hermite.cc:
```diff
@@ -0,0 +1,840 @@
+      const unsigned int dofs          = fe_herm.n_dofs_per_cell();
+      const unsigned int regularity    = fe_herm.get_regularity();
+      const unsigned int dofs_per_dim  = 2 * regularity + 2;
+      const unsigned int dofs_per_quad = dofs_per_dim * dofs_per_dim;
```
```suggestion
      const unsigned int n_dofs_per_quad = dofs_per_dim * dofs_per_dim;

```

peterrum commented on source/fe/fe_hermite.cc:
```diff
@@ -0,0 +1,840 @@
+      FiniteElementData<dim>(internal::get_hermite_dpo_vector(dim, reg),
+                             1,
+                             2 * reg + 1,
+                             (reg ? FiniteElementData<dim>::H2 :
```
```suggestion
                             (reg > 0 ? FiniteElementData<dim>::H2 :

```

peterrum commented on source/fe/fe_hermite.cc:
```diff
@@ -0,0 +1,840 @@
+
+// Constructors
+template <int dim, int spacedim>
+FE_Hermite<dim, spacedim>::FE_Hermite(const unsigned int reg)
```
```suggestion
FE_Hermite<dim, spacedim>::FE_Hermite(const unsigned int regularity)

```

peterrum commented on source/fe/fe_hermite.cc:
```diff
@@ -0,0 +1,840 @@
+std::vector<unsigned int>
+FE_Hermite<dim, spacedim>::get_lexicographic_to_hierarchic_numbering() const
+{
+  return internal::hermite_lexicographic_to_hierarchic_numbering<dim>(
+    this->regularity);
```
What about storing this information internally?

peterrum commented on source/fe/fe_hermite.cc:
```diff
@@ -0,0 +1,840 @@
+
+
+
+/*
+ * A special version of this function is needed for Hermite elements,
+ * due to the need to rescale different shape functions according to
+ * the derivative order they represent.
+ */
```
```suggestion
```
This is already stated in the header.

peterrum commented on source/fe/fe_hermite.cc:
```diff
@@ -0,0 +1,840 @@
+      (dynamic_cast<const typename MappingCartesian<dim>::InternalData *>(
+         &mapping_internal) != nullptr),
+    ExcInternalError()); /*
+const typename MappingHermite<dim, spacedim>::InternalData
+&mapping_internal_herm = static_cast<const typename MappingHermite<dim,
+spacedim>::InternalData &>( mapping_internal);*/
```
```suggestion
```

peterrum commented on source/fe/fe_hermite.cc:
```diff
@@ -0,0 +1,840 @@
+      fe_internal);
+
+  Assert(
+    (dynamic_cast<const typename MappingHermite<dim, spacedim>::InternalData *>(
+       &mapping_internal) != nullptr) ||
```
Let's keep this simpler and not think about a mapping class that might be implemented in the future.

peterrum commented on source/fe/fe_hermite.cc:
```diff
@@ -0,0 +1,840 @@
+  if ((flags & update_values) &&
+      (cell_similarity != CellSimilarity::translation))
+    {
+      internal::Rescaler<dim, spacedim, double> shape_fix;
+      for (unsigned int i = 0; i < output_data.shape_values.size(0); ++i)
+        for (unsigned int q = 0; q < output_data.shape_values.size(1); ++q)
+          output_data.shape_values(i, q) = fe_data.shape_values(i, q);
+      shape_fix.rescale_fe_hermite_values(shape_fix,
+                                          *this,
+                                          mapping_internal, //_herm,
+                                          output_data.shape_values);
```
Out of curiosity, couldn't one call `FE::Poly::fill_fe_values()` and to the rescaling as postprocessing? I know these lines of code from somewhere.

peterrum commented on source/fe/fe_hermite.cc:
```diff
@@ -0,0 +1,840 @@
+     &mapping_internal_herm = static_cast<const typename
+     MappingHermite<dim,spacedim>::InternalData &>( mapping_internal);*/
+
+  AssertDimension(quadrature.size(), 1U);
```
```suggestion
  AssertDimension(quadrature.size(), 1);
```
Did you get compiler warning?

peterrum commented on source/fe/fe_hermite.cc:
```diff
@@ -0,0 +1,840 @@
+      (dynamic_cast<const typename MappingCartesian<dim, spacedim>::InternalData
+                      *>(&mapping_internal) != nullptr),
+    ExcInternalError());
+  /*const typename MappingHermite<dim, spacedim>::InternalData
+     &mapping_internal_herm = static_cast<const typename
+     MappingHermite<dim,spacedim>::InternalData &>( mapping_internal);*/
```
```suggestion
```

peterrum commented on include/deal.II/numerics/vector_tools_hermite.templates.h:
```diff
@@ -0,0 +1,630 @@
+
+namespace VectorTools
+{
+  // Internal namespace for implement Hermite boundary projection methods
```
```suggestion
```

peterrum replied:
It looks like you missed my comments in this file.

peterrum commented on include/deal.II/numerics/vector_tools_hermite.templates.h:
```diff
@@ -0,0 +1,630 @@
+       * 4 or six faces of reference cell
+       */
+      const unsigned int degree = dof_handler.get_fe().degree;
+      const unsigned int regularity =
+        dynamic_cast<const FE_Hermite<dim> &>(dof_handler.get_fe())
+          .get_regularity();
```
```suggestion
      const auto fe = dynamic_cast<const FE_Hermite<dim> * >(&dof_handler.get_fe());
      Assert(fe);
      const unsigned int regularity =fe->get_regularity();
```

peterrum commented on include/deal.II/numerics/vector_tools_hermite.templates.h:
```diff
@@ -0,0 +1,630 @@
+            Assert(index < dofs_per_face,
+                   ExcDimensionMismatch(index, dofs_per_face));
+
+            test_cell->face(2 * d)->get_dof_indices(
+              dofs_on_face, test_cell->active_fe_index());
```
Where is `dofs_on_face` used?

peterrum commented on include/deal.II/numerics/vector_tools_hermite.templates.h:
```diff
@@ -0,0 +1,630 @@
+  {
+    template <int dim, int spacedim = dim, typename Number = double>
+    void
+    get_constrained_hermite_boundary_dofs(
```
Could you add a comment what this function does?

peterrum commented on include/deal.II/numerics/vector_tools_hermite.templates.h:
```diff
@@ -0,0 +1,630 @@
+                           ExcDimensionMismatch(index, dof_to_boundary.size()));
+
+                    if (dof_to_boundary[index] == numbers::invalid_dof_index)
+                      dof_to_boundary[index] = no_constrained_dofs++;
```
```suggestion
                      dof_to_boundary[index] = n_constrained_dofs++;
```

peterrum commented on include/deal.II/numerics/vector_tools_hermite.templates.h:
```diff
@@ -0,0 +1,630 @@
+
+
+
+    template <int dim, int spacedim = dim, typename Number = double>
```
```suggestion
    template <int dim, int spacedim, typename Number>
```

peterrum commented on include/deal.II/numerics/vector_tools_hermite.templates.h:
```diff
@@ -0,0 +1,630 @@
+
+    template <int dim, int spacedim = dim, typename Number = double>
+    void
+    do_hermite_direct_projection(
```
Could you add a comment?

peterrum commented on include/deal.II/numerics/vector_tools_hermite.templates.h:
```diff
@@ -0,0 +1,630 @@
+                       .find(cell->face(direction)->boundary_id())
+                       ->second;
+
+                  const FiniteElement<dim, spacedim> &fe_herm =
```
```suggestion
                  const FiniteElement<dim, spacedim> &fe =
```

peterrum commented on include/deal.II/numerics/vector_tools_hermite.templates.h:
```diff
@@ -0,0 +1,630 @@
+          Assert(component_mapping.size() == 0, ExcNotImplemented());
+
+          for (const auto &cell : dof_handler.active_cell_iterators())
+            for (const unsigned int direction :
+                 GeometryInfo<dim>::face_indices())
```
```suggestion
            for (const unsigned int face : cell->face_indices())
```

peterrum commented on include/deal.II/numerics/vector_tools_hermite.templates.h:
```diff
@@ -0,0 +1,630 @@
+                nullptr),
+             ExcInternalError());
+
+      // For dim=1, the problem simplifies to interpolation at the boundaries
```
Why is this the case?

peterrum commented on include/deal.II/numerics/vector_tools_hermite.templates.h:
```diff
@@ -0,0 +1,630 @@
+                                            dof_to_boundary_mapping,
+                                            next_boundary_index);
+
+      SparsityPattern sparsity;
```
This is a serial code. What are the plans regarding parallelisation?

peterrum commented on include/deal.II/numerics/vector_tools_hermite.templates.h:
```diff
@@ -0,0 +1,630 @@
+      }
+
+      // Assert mesh is not partially refined
+      // TODO: Add functionality on partially refined meshes
```
Could you add this to the issue.

peterrum commented on include/deal.II/numerics/vector_tools_hermite.templates.h:
```diff
@@ -0,0 +1,630 @@
+
+      // Assert mesh is not partially refined
+      // TODO: Add functionality on partially refined meshes
+      int level = -1;
```
```suggestion
      usigned int level = numbers::invalid_unsigned_int;
```

peterrum commented on include/deal.II/numerics/vector_tools_hermite.templates.h:
```diff
@@ -0,0 +1,630 @@
+           * Current implementation uses a direct solver to better account for
+           * the mass matrix being potentially sufficiently ill-conditioned to
+           * prevent iterations from converging.
+           * TODO: Find a good preconditioner to allow more efficient iterative
+           * methods
```
TODO.

peterrum commented on include/deal.II/numerics/vector_tools_hermite.templates.h:
```diff
@@ -0,0 +1,630 @@
+             * DOFs that are actually constrained, is applied to the boundary
+             * DOFs in the global solution. Remember: i is the global dof
+             * number, dof_to_boundary_mapping[i] is the number on the boundary
+             * and thus in the solution vector
```
```suggestion
             * and thus in the solution vector.
```

peterrum commented on include/deal.II/numerics/vector_tools_hermite.templates.h:
```diff
@@ -0,0 +1,630 @@
+    std::vector<unsigned int>                  component_mapping)
+  {
+    /*
+     * This version implements projected values directly,
```
What does `directly` mean?

peterrum commented on include/deal.II/numerics/vector_tools_hermite.templates.h:
```diff
@@ -0,0 +1,630 @@
+            nullptr) ||
+             (dynamic_cast<const MappingCartesian<dim> *>(&mapping_h) !=
+              nullptr),
+           ExcMessage("project_hermite should only be used with a mapping "
+                      "compatible with Hermite finite element methods."));
```
```suggestion
           ExcMessage("The function project_hermite() can only be used with a mapping "
                      "compatible with Hermite finite element methods."));
```

peterrum commented on include/deal.II/numerics/vector_tools_hermite.templates.h:
```diff
@@ -0,0 +1,630 @@
+
+    if (enforce_zero_boundary)
+      {
+        std::map<types::boundary_id, const Function<spacedim, number> *>
```
Add `assert(project_to_boundary_first==false)`

peterrum reviewed (commented):
The rest. It looks very good! I will take another look at include/deal.II/numerics/vector_tools_hermite.templates.h once you have added some comments.

peterrum commented:
@DanielPaukner Question: I know `MappingCaresian` is not enough for you, but is could you this to solve one timestep or so? In this case the deformation should be still zero. I.e., it would not make any difference.

DanielPaukner commented:
@peterrum True, I will give it a try and get back to you!

ivweber commented on source/base/polynomials_hermite.cc:
```diff
@@ -0,0 +1,162 @@
+      ExcMessage(
+        "The value passed to regularity is too high, this may be due to the requested value being too high for numerical stability, or due to a bug in the code."));
+
+    const unsigned int sz = regularity + 1;
```
sz stands for size, it's the size of the for loops. But yes, it is not necessary since <= can be used with regularity in the loop checks instead

ivweber commented on source/base/polynomials_hermite.cc:
```diff
@@ -0,0 +1,162 @@
+    int                mult;
+
+    coeffs[0] = 1;
+    for (unsigned int i = 1; i < sz; ++i)
```
I will look into this. I was advised to avoid using ints instead of unsigned ints in for loops, but this could resolve the issues.

ivweber commented on include/deal.II/fe/fe_hermite.h:
```diff
@@ -0,0 +1,257 @@
+  // Other functions
+  /**
+   * Returns a @p string that uniquely identifies the finite element object, in
+   * this case <code>FE_Hermite<dim,spacedim>(regularity)</code> with @p dim,
+   * @p spacedim and @p regularity replaced with the correct values.
```
I've just checked the guide to writing documentation:
https://www.dealii.org/developer/developers/writing-documentation.html
Apparently `@p` is used to set only the following word to typewriter font and <code></code> is recommended for longer sections of text.

ivweber replied:
That should have read "and `<code></code>` is recommended..."

ivweber commented on include/deal.II/fe/fe_hermite.h:
```diff
@@ -0,0 +1,257 @@
+   * guarantee rectangular cells which ensures that the directions of
+   * derivatives does not change. The derivative continuity can then be
+   * enforced by re-scaling shape functions so that the magnitude of each
+   * derivative is equivalent to the derivative on the reference interval. This
+   * re-scaling is implemented by an internal object called
+   * <code>Rescaler<dim,spacedim></code>.
```
Gradients, Hessians, 3rd derivatives... need to be re-scaled for existing finite elements, and this is usually handled by `FiniteElement::fill_fe_values()` or `FiniteElement::fill_fe_face_values()`, but using data from the mapping object.

ivweber commented on include/deal.II/numerics/vector_tools_hermite.templates.h:
```diff
@@ -0,0 +1,630 @@
+ * Hermite, which is why this name is reserved here.
+ */
+
+template <int dim, int spacedim = dim>
+using MappingHermite = MappingCartesian<dim>;
```
> One question: I thought you had to adjust the mapping class. Is this part of this PR?

No, the current implementation of `MappingCartesian` works with Hermite elements.

ivweber commented on source/fe/fe_hermite.cc:
```diff
@@ -0,0 +1,840 @@
+
+
+
+  template <int xdim, int xspacedim = xdim, typename xNumber = double>
```
Different values of `xdim` are defined below separately, but `spacedim` and `Number` are kept as template arguments. An extra `x` is added to template variables here to prevent compilation errors.

ivweber commented on source/fe/fe_hermite.cc:
```diff
@@ -0,0 +1,840 @@
+
+  template <>
+  void
+  hermite_hierarchic_to_lexicographic_numbering<4>(
```
I might have been too ambitious here. I've taken it out since it never gets instantiated by fe_hermite.inst.in.

peterrum commented:
Let me answer this way to your comments. You have "resolved" the discussions above so that I cannot simply jump to the right discussion to add the comment there.

> Degree 13 was the highest we could go in one dimension and still get usable results when we were investigating time-step restrictions of Hermite elements with the wave equation. In theory higher degrees can be defined, but at the moment they're not practically usable.

I would reformulate the statement in this case.

> I've just checked the guide to writing documentation:
https://www.dealii.org/developer/developers/writing-documentation.html
Apparently @p is used to set only the following word to typewriter font and is recommended for longer sections of text.

OK!

> Different values of xdim are defined below separately, but spacedim and Number are kept as template arguments. An extra x is added to template variables here to prevent compilation errors.

Does the class actually needs to be templated? As it is right now, one does not need a class. You could convert it to a namespace. Are you planing to add fields to the class in the future?

kronbichler commented:
@ivweber and I went through the code and re-based. I think most of the comments were addressed, and I saw some things that we might want to fix some follow-up PRs. Ready for a second review.

kronbichler commented:
@ivweber can you have a look at the compile error shown by the continuous integration test
```
[](https://jenkins.tjhei.info/blue/organizations/jenkins/dealii/detail/PR-13239/27/pipeline#step-85-log-1389)[](https://jenkins.tjhei.info/blue/organizations/jenkins/dealii/detail/PR-13239/27/pipeline#step-85-log-1390)[](https://jenkins.tjhei.info/blue/organizations/jenkins/dealii/detail/PR-13239/27/pipeline#step-85-log-1391)[](https://jenkins.tjhei.info/blue/organizations/jenkins/dealii/detail/PR-13239/27/pipeline#step-85-log-1392)[](https://jenkins.tjhei.info/blue/organizations/jenkins/dealii/detail/PR-13239/27/pipeline#step-85-log-1393)FAILED: /usr/bin/mpicxx   -DBOOST_NO_AUTO_PTR -DDEBUG -Isource/fe -Iinclude -I/jenkins/workspace/dealii_PR-13239/include -I/jenkins/workspace/dealii_PR-13239/bundled/tbb-2018_U2/include -I/jenkins/workspace/dealii_PR-13239/bundled/boost-1.70.0/include -I/jenkins/workspace/dealii_PR-13239/bundled/umfpack/UMFPACK/Include -I/jenkins/workspace/dealii_PR-13239/bundled/umfpack/AMD/Include -I/jenkins/workspace/dealii_PR-13239/bundled/muparser_v2_2_4/include -isystem /usr/include/mpich -isystem /home/dealii/libs/parmetis-4.0.3/include -isystem /home/dealii/libs/trilinos-release-12-10-1/include -isystem /home/dealii/libs/superlu_dist_5.1.2/include -isystem /home/dealii/libs/petsc-3.7.6/include -isystem /home/dealii/libs/adolc-2.6.4-rc1/include -isystem /home/dealii/libs/assimp-4.1.0/include -isystem /home/dealii/libs/oce-OCE-0.18.2/include/oce -isystem /home/dealii/libs/p4est-2.0/FAST/include -isystem /home/dealii/libs/slepc-3.7.3/include -isystem /home/dealii/libs/sundials-3.1.0/include -pedantic -fPIC -Wall -Wextra -Wmissing-braces -Woverloaded-virtual -Wpointer-arith -Wsign-compare -Wsuggest-override -Wswitch -Wsynth -Wwrite-strings -Wno-deprecated-declarations -Wno-literal-suffix -Wno-psabi -fopenmp-simd -Werror -std=c++14 -Wno-parentheses -Wno-unused-local-typedefs -O0 -ggdb -Wa,--compress-debug-sections -Og -MMD -MT source/fe/CMakeFiles/obj_fe_debug.dir/fe_values.cc.o -MF source/fe/CMakeFiles/obj_fe_debug.dir/fe_values.cc.o.d -o source/fe/CMakeFiles/obj_fe_debug.dir/fe_values.cc.o -c /jenkins/workspace/dealii_PR-13239/source/fe/fe_values.cc

/jenkins/workspace/dealii_PR-13239/source/fe/fe_values.cc:2620:33: fatal error: fe_values.decl.1.inst: No such file or directory

compilation terminated.

FAILED: /usr/bin/mpicxx   -DBOOST_NO_AUTO_PTR -DDEBUG -Isource/fe -Iinclude -I/jenkins/workspace/dealii_PR-13239/include -I/jenkins/workspace/dealii_PR-13239/bundled/tbb-2018_U2/include -I/jenkins/workspace/dealii_PR-13239/bundled/boost-1.70.0/include -I/jenkins/workspace/dealii_PR-13239/bundled/umfpack/UMFPACK/Include -I/jenkins/workspace/dealii_PR-13239/bundled/umfpack/AMD/Include -I/jenkins/workspace/dealii_PR-13239/bundled/muparser_v2_2_4/include -isystem /usr/include/mpich -isystem /home/dealii/libs/parmetis-4.0.3/include -isystem /home/dealii/libs/trilinos-release-12-10-1/include -isystem /home/dealii/libs/superlu_dist_5.1.2/include -isystem /home/dealii/libs/petsc-3.7.6/include -isystem /home/dealii/libs/adolc-2.6.4-rc1/include -isystem /home/dealii/libs/assimp-4.1.0/include -isystem /home/dealii/libs/oce-OCE-0.18.2/include/oce -isystem /home/dealii/libs/p4est-2.0/FAST/include -isystem /home/dealii/libs/slepc-3.7.3/include -isystem /home/dealii/libs/sundials-3.1.0/include -pedantic -fPIC -Wall -Wextra -Wmissing-braces -Woverloaded-virtual -Wpointer-arith -Wsign-compare -Wsuggest-override -Wswitch -Wsynth -Wwrite-strings -Wno-deprecated-declarations -Wno-literal-suffix -Wno-psabi -fopenmp-simd -Werror -std=c++14 -Wno-parentheses -Wno-unused-local-typedefs -O0 -ggdb -Wa,--compress-debug-sections -Og -MMD -MT source/fe/CMakeFiles/obj_fe_debug.dir/fe_values_inst2.cc.o -MF source/fe/CMakeFiles/obj_fe_debug.dir/fe_values_inst2.cc.o.d -o source/fe/CMakeFiles/obj_fe_debug.dir/fe_values_inst2.cc.o -c /jenkins/workspace/dealii_PR-13239/source/fe/fe_values_inst2.cc

In file included from /jenkins/workspace/dealii_PR-13239/source/fe/fe_values_inst2.cc:20:0:

/jenkins/workspace/dealii_PR-13239/source/fe/fe_values.cc:2620:33: fatal error: fe_values.decl.1.inst: No such file or directory
```
I think we made a mistake when rebasing/squashing the file `source/fe/fe_values.cc` and got some intermediate state from that, but this file should not have any change in this PR. Can you please pick the file from the current master version of deal.II to replace the content on this branch, and force-push again?

kronbichler reviewed (approved):
From my point of view, this looks good, but we need someone else to also look at the code.

kronbichler commented:
/rebuild

peterrum commented:
I'll have at then end of the week.

peterrum commented on include/deal.II/base/polynomials_hermite.h:
```diff
@@ -0,0 +1,157 @@
+namespace Polynomials
+{
+  /**
+   * This class implements Hermite interpolation polynomials
```
Could you add a reference?

ivweber replied:
Yes, of course. Is there a standard format for adding academic references to deal.II documentation?

peterrum replied:
E.g.: 

https://github.com/dealii/dealii/blob/e6779913e7f770652da442b3c2460dec3e864029/include/deal.II/base/mpi_consensus_algorithms.h#L316-L317

https://github.com/dealii/dealii/blob/8e3873780d4d35d18dfaf7c2a5364e39aa175402/doc/doxygen/references.bib#L1684-L1694

peterrum commented on include/deal.II/base/polynomials_hermite.h:
```diff
@@ -0,0 +1,157 @@
+    /**
+     * Degree of the polynomial basis being used.
+     */
+    unsigned int degree;
+
+    /**
+     * The order of the highest derivative in which the Hermite
+     * basis can be used to impose continuity across element
+     * boundaries. It's related to the degree $p$ by
+     * $p = 2 \times\mathtt{regularity} +1$.
+     */
+    unsigned int regularity;
+
+    /**
+     * This variable stores the derivative that the shape function
+     * corresponds to at the element boundary given by <code>side</code>.
+     */
+    unsigned int side_index;
+
+    /**
+     * This stores whether the shape function corresponds to a non-zero
+     * value or derivative at $x=0$ on the reference interval
+     * ($\mathtt{side} =0$), or at $x=1$ ($\mathtt{side} =1$).
+     */
+    unsigned int side;
```
These could be const, couldn't they?

peterrum commented on include/deal.II/matrix_free/shape_info.templates.h:
```diff
@@ -1095,7 +1096,9 @@ namespace internal
 
           // then check if the base element is supported or not
-          if (dynamic_cast<const FE_Poly<dim, spacedim> *>(fe_ptr) != nullptr)
+          if (dynamic_cast<const FE_Poly<dim, spacedim> *>(fe_ptr) != nullptr &&
+              dynamic_cast<const FE_Hermite<dim, spacedim> *>(fe_ptr) ==
+                nullptr)
```
Is this used somewhere?

peterrum commented on source/dofs/dof_tools_sparsity.cc:
```diff
@@ -504,10 +506,18 @@ namespace DoFTools
 
             // make sparsity pattern for this cell
+            /*
+             * This loop has been edited to check for invalid dof indices, which
```
?

peterrum commented on include/deal.II/fe/fe_hermite.h:
```diff
@@ -0,0 +1,258 @@
+
+DEAL_II_NAMESPACE_CLOSE
+
+#include <deal.II/numerics/vector_tools_hermite.h>
```
```suggestion
```
What is the reason for having this include here?

ivweber replied:
I think this was left over from when the projection functions were declared in fe_hermite to avoid changing any code that used those functions. I will remove this.

peterrum commented on include/deal.II/fe/fe_hermite.h:
```diff
@@ -0,0 +1,258 @@
+
+
+/**
+ * This class implements a Hermite basis of maximum regularity elements.
```
Could you add reference here as well?

peterrum commented on include/deal.II/fe/fe_hermite.h:
```diff
@@ -0,0 +1,258 @@
+   * boundaries, it is necessary to account for any changes to derivatives
+   * from those on the reference element due to the current cell mapping.
+   *
+   * At present this is done by only allowing certain cell mappings that
```
Could you list somewhere prominently the supported mapping classes?

peterrum commented on include/deal.II/fe/fe_hermite.h:
```diff
@@ -0,0 +1,258 @@
+                                                                       spacedim>
+      &output_data) const override;
+
+  using FiniteElement<dim, spacedim>::fill_fe_face_values;
```
Why is this line needed?

ivweber replied:
If I recall why this was added correctly, before this line was added the code automatically used the pre-existing function for Lagrange elements, and ignored the new version that we needed for correct results with Hermite elements.

peterrum commented on include/deal.II/fe/fe_hermite.h:
```diff
@@ -0,0 +1,258 @@
+   * Wrapper function for FE_Poly::get_data() that ensures relevant
+   * shape value data is copied to the new data object as well.
+   */
+  virtual std::unique_ptr<
+    typename FiniteElement<dim, spacedim>::InternalDataBase>
+  get_data(
```
Could you split up declaration and definition?

peterrum commented on include/deal.II/fe/fe_hermite.h:
```diff
@@ -0,0 +1,258 @@
+   * Returns the regularity of the Hermite FE basis.
+   */
+  inline unsigned int
+  get_regularity() const
+  {
+    return this->regularity;
+  };
```
Same here?

peterrum commented on include/deal.II/fe/fe_hermite.h:
```diff
@@ -0,0 +1,258 @@
+
+
+
+public:
```
Could you sort the function so that public functions are in one block?

peterrum commented on include/deal.II/fe/fe_hermite.h:
```diff
@@ -0,0 +1,258 @@
+   * $\frac{d^{2}f}{dx \; dy}$ would be counted as a first order derivative
+   * of $f$, as an example.
+   */
+  unsigned int regularity;
```
Could this be const?

peterrum commented on source/fe/fe_hermite.cc:
```diff
@@ -0,0 +1,802 @@
+
+
+
+/*
+ * In the future it may be possible to relax the restrictions on mappings
+ * that can be used with FE_Hermite and implement these in a mapping
+ * class, so the name MappingHermite is used in some places. At the moment
+ * this has not been investigated properly, so MappingHermite is defined
+ * here as an alias for MappingCartesian.
+ */
+template <int dim, int spacedim = dim>
+using MappingHermite = MappingCartesian<dim>;
```
I would remove this if you don't have concrete plans to work on this topic.

peterrum commented on source/fe/fe_hermite.cc:
```diff
@@ -0,0 +1,802 @@
+  template <int dim>
+  void
+  hermite_hierarchic_to_lexicographic_numbering(const unsigned int regularity,
+                                                std::vector<unsigned int> &h2l);
```
```suggestion
                                                std::vector<unsigned int> &hierarchic_to_lexicographic_numbering);
```

peterrum commented on source/fe/fe_hermite.cc:
```diff
@@ -0,0 +1,802 @@
+          output_data.shape_values(i, q) = fe_data.shape_values(i, q);
+      shape_fix.rescale_fe_hermite_values(shape_fix,
+                                          *this,
+                                          mapping_internal, //_herm,
```
```suggestion
                                          mapping_internal,
```

peterrum commented on source/fe/fe_hermite.cc:
```diff
@@ -0,0 +1,802 @@
+      internal::Rescaler<dim, spacedim, Tensor<1, spacedim>> grad_fix;
+      grad_fix.rescale_fe_hermite_values(grad_fix,
+                                         *this,
+                                         mapping_internal, //_herm,
```
```suggestion
                                         mapping_internal,
```

same a couple of times below

peterrum reviewed (commented):
Close! Thanks for the modifications!

@drwells The PR is already in a very good shape. If you still want to review, now it is the time!

peterrum commented:
@ivweber Could you also add an change-log entry here: https://github.com/dealii/dealii/tree/master/doc/news/changes/major

drwells reviewed (changes requested):
I have a few general comments - I still need to read through the proper C^1 implementation.

There are still a lot of unresolved comments from previous reviews (particularly from Peter). I think we should address those before continuing. It would also help (given how big this patch is) if we split some of it out into little PRs.

ivweber commented:
> I'd like to get this in the library. @ivweber Do you have time to work through some of the review comments? If you need help addressing any of them let us know! I'd be happy to push a patch or two if you think that would help.

Apologies! I was busy earlier in the year with coursework and preparing for a conference, so this got side-lined as I could not see any chance of resolving this quickly. I will start looking through the requirements again for this pull request. Is there an easy way to see only the unresolved issues? I am not very familiar with GitHub and have basically been learning version control as I go.

ivweber commented on source/base/polynomials_hermite.cc:
```diff
@@ -0,0 +1,128 @@
+     * ints. At present static casts to int are used to fix the issue when
+     * needed. If this bug occurs it usually affects FE_Hermite(2) and higher,
+     * with shape values taking large magnitude values in the interval (0,1).
+     *  - Ivy Weber
```
Yes, there were a couple of places where an intermediate value in calculations could be negative even when the final value was guaranteed to be positive, and this caused errors. I will change this to inline comments.

drwells commented:
Not a problem - I'm glad you're back to finish it now.

> Is there an easy way to see only the unresolved issues?

Unfortunately there is not - this is a part of GitHub's interface that's really lacking. It doesn't work well with large PRs like this one - 213 comments is just more than it can reasonably handle. The best workaround we have is to divide PRs into smaller PRs. One possibility would be to put the Polynomial class and some tests for it in a first PR, the FE in the second, and VectorTools additions in the third.

ivweber commented:
> The best workaround we have is to divide PRs into smaller PRs. One possibility would be to put the Polynomial class and some tests for it in a first PR, the FE in the second, and VectorTools additions in the third.

I think this is a good idea. I will start prioritising the comments with the Polynomial class. Would the easiest way to do this be to open a different PR for just the polynomials?

peterrum commented:
@ivweber I think there only a few features remaining (`VectorTools`, ...) we need to merge now that #15415 is in master!

ivweber commented:
Is there a way to see more output about test failures from Jenkins? The tests pass without issue on my laptop, so I can't see what is causing the serial test to fail.

ivweber commented:
@drwells I've re-opened this pull request now that PolynomialsHermite and FE_Hermite have been merged via separate pull requests, as suggested. I have made some of the requested changes to the VectorTools addition, but I have left the names of functions largely unchanged to avoid confusion, as some of the boundary methods make use of knowledge about the structure of Hermite elements. 
I haven't had time yet to look into modifying the functions to use AffineConstraints instead of maps, and I can start looking into this if this is a hard requirement for adding the code to the library.

peterrum commented on doc/news/changes/major/20230919IvyWeber:
```diff
@@ -0,0 +1,3 @@
+New: Added methods for enforcing boundary conditions directly with Hermite elements on an aligned rectangular grid.
+<br>
+(Ivy Weber, 2023/09/19)
```
I guess this is not needed anymore?

peterrum commented on include/deal.II/base/types.h:
```diff
@@ -233,6 +233,14 @@ namespace numbers
+  /**
+   * A degree of freedom index used by the implementation of Hermite finite
+   * elements to denote functions that are assigned on the boundary of the
+   * domain, but that should be treated as internal domain d.o.f.s.
```
```suggestion
   * domain, but that should be treated as internal domain DoFs.
```

peterrum commented on include/deal.II/numerics/vector_tools_hermite.h:
```diff
@@ -0,0 +1,133 @@
+   */
+  template <int dim, int spacedim = dim, typename Number = double>
+  void
+  project_hermite_boundary_values(
```
This has not been addressed yet.

peterrum commented on source/dofs/dof_tools_sparsity.cc:
```diff
@@ -511,14 +511,18 @@ namespace DoFTools
             cols.clear();
             for (const auto &dof : dofs_on_this_face)
-              cols.push_back(dof_to_boundary_mapping[dof]);
+              if (dof_to_boundary_mapping[dof] !=
+                  numbers::unconstrained_boundary_dof_index)
+                cols.push_back(dof_to_boundary_mapping[dof]);
```
This looks like an unrelated change. How did this become relevant?

peterrum commented on include/deal.II/numerics/vector_tools_hermite.templates.h:
```diff
@@ -0,0 +1,597 @@
+// ---------------------------------------------------------------------
```
The comments in include/deal.II/numerics/vector_tools_hermite.templates.h seem to be not addressed, yet.

peterrum reviewed (commented):
@ivweber There some unaddressed comments by @drwells  and by me. I don't think it is too much. Once, you have addressed them, I think we are good to merge!

ivweber commented:
I have closed this for now, as I have several other deadlines at Uppsala that I need to prioritise and Hermite elements appear to work with immersed boundaries anyway. I will come back to this at a future date when I have the time to properly address everything.
